src/mesh.f90

```diff
@@ -0,0 +1,375 @@
+
+    integer, dimension(3), private :: vert_dims_padded ! local domain size including padding (cartesian structure)
+    integer, dimension(3), private :: vert_dims ! local number of vertices in each direction without padding (cartesian structure)
+    integer, dimension(3), private :: cell_dims ! local number of cells in each direction without padding (cartesian structure)
+    logical, dimension(3), private :: periodic_dir ! Whether or not a direction has a periodic BC
+    integer, private :: sz
+    class(geo_t), allocatable :: geo ! object containing geometry information
+    class(parallel_t), allocatable :: par ! object containing parallel domain decomposition information
+  contains
+    procedure :: get_SZ
+
+    procedure :: get_n_groups_dir
+    procedure :: get_n_groups_phi
+    generic :: get_n_groups => get_n_groups_dir, get_n_groups_phi
+
+    procedure :: get_dims_dir
+    procedure :: get_dims_phi
+    procedure :: get_dims_phi_dataloc
+    generic :: get_dims => get_dims_dir, get_dims_phi, get_dims_phi_dataloc
+
+    procedure :: get_n_dir 
+    procedure :: get_n_phi
+    generic :: get_n => get_n_dir, get_n_phi
+
+    procedure :: get_padded_dims_phi
+    procedure :: get_padded_dims_dir
+    generic :: get_padded_dims => get_padded_dims_dir, get_padded_dims_phi
+  
+    procedure :: get_coordinates
+
+    procedure :: set_sz
+    procedure :: set_padded_dims
+  end type mesh_t
+
+  interface mesh_t
+    module procedure mesh_init
+  end interface mesh_t
+
+  contains 
+
+  function mesh_init(dims_global, nproc_dir, L_global) result(mesh)
+    !! Completely initialise the mesh object. 
+    !! Upon initialisation the mesh object can be read-only and shouldn't be edited
+    !! Takes as argument global information about the mesh like its length, number of cells and decomposition in each direction
+    integer, dimension(3), intent(in) :: dims_global
+    integer, dimension(3), intent(in) :: nproc_dir ! Number of proc in each direction
+    real(dp), dimension(3), intent(in) :: L_global
+    type(mesh_t) :: mesh
+
+    integer :: nx, ny, nz, dir
+    integer :: ierr
+    logical :: is_last_domain
+
+    allocate(mesh%geo)
+    allocate(mesh%par)
+    mesh%dims_global(:) = dims_global
+    ! Geometry
+    mesh%geo%L = L_global
+    mesh%geo%d = mesh%geo%L(:) / mesh%dims_global(:)
+
+    ! Parallel domain decomposition
+    mesh%par%nproc_dir(:) = nproc_dir
+    mesh%par%nproc = product(nproc_dir(:))
+    call MPI_Comm_rank(MPI_COMM_WORLD, mesh%par%nrank, ierr)
+    call MPI_Comm_size(MPI_COMM_WORLD, mesh%par%nproc, ierr)
+    call domain_decomposition(mesh)
+
+    ! Local mesh dimensions
+    nx = mesh%dims_global(1)/mesh%par%nproc_dir(1)
+    ny = mesh%dims_global(2)/mesh%par%nproc_dir(2)
+    nz = mesh%dims_global(3)/mesh%par%nproc_dir(3)
+
+
+    ! TODO: fixme: hard code no periodic boundary (for now)
+    mesh%periodic_dir(:) = .false.
```

Comment:
if you're not planning to fix in the current PR you can set this to periodic so that the TGV case works once this is merged

Reply:
Implemented it and set default to `.true.`